Add integer multiplication methods for Seq, UnknownSeq and MutableSeq

chris-rands opened the pull request:
This pull request addresses issue #1659

The issue explains the rationale, but here is a small example of the utility of the new methods:

```
>>> from Bio.Seq import Seq
>>> Seq('AT') * 2
Seq('ATAT', Alphabet())
>>> 2 * Seq('AT')    
Seq('ATAT', Alphabet())
>>> seq = Seq('AT')
>>> seq *= 2
>>> seq
Seq('ATAT', Alphabet())
```

The methods should work with all the `Seq`, `UnknownSeq`, and `MutableSeq` objects and with any alphabet. I also implemented some unit tests.

Let me know if this looks ok?

I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

I have read the ``CONTRIBUTING.rst`` file and understand that AppVeyor and
TravisCI will be used to confirm the Biopython unit tests and ``flake8`` style
checks pass with these changes.

I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

peterjc commented on Bio/Seq.py:
```diff
@@ -345,6 +345,54 @@ def __radd__(self, other):
+        return self.__class__(str(self) * other, self.alphabet)
+
+    def __imul__(self, other):
+        """Multiply Seq in-place
```
Needs trailing full stop, see PEP257 https://www.python.org/dev/peps/pep-0257/

peterjc commented on Bio/Seq.py:
```diff
@@ -345,6 +345,54 @@ def __radd__(self, other):
+        """Multiply Seq in-place
+
+        Note although Seq is immutable, the in-place method is
+        included to match the behaviour for regular Python strings
```
Can you close that sentence with a full stop please. Likewise for the other docstrings.

peterjc commented on Bio/Seq.py:
```diff
@@ -1953,6 +2049,57 @@ def __radd__(self, other):
+
+        if not isinstance(other, int):
+            raise TypeError("can't multiply MutableSeq by non-int type")
+        return self.__class__(str(self) * other, self.alphabet)
```
I wonder if we can make this much more efficient by working directly with the ``self.data`` array, avoiding conversion to and from a string?

peterjc commented on Bio/Seq.py:
```diff
@@ -345,6 +345,54 @@ def __radd__(self, other):
+        >>> Seq('ATG', generic_dna) * 2
+        Seq('ATGATG', DNAAlphabet())
+        """
+
```
Please remove this blank line, likewise for the other methods.

peterjc commented:
The TravisCI check is failing on some fairly minor style issues:

```
$ flake8 Bio/
Bio/Seq.py:348:1: D202 No blank lines allowed after function docstring
Bio/Seq.py:348:1: D400 First line should end with a period
Bio/Seq.py:363:1: D202 No blank lines allowed after function docstring
Bio/Seq.py:363:1: D400 First line should end with a period
Bio/Seq.py:378:1: D202 No blank lines allowed after function docstring
Bio/Seq.py:378:1: D400 First line should end with a period
Bio/Seq.py:1396:1: D202 No blank lines allowed after function docstring
Bio/Seq.py:1396:1: D400 First line should end with a period
Bio/Seq.py:1411:1: D202 No blank lines allowed after function docstring
Bio/Seq.py:1411:1: D400 First line should end with a period
Bio/Seq.py:1426:1: D202 No blank lines allowed after function docstring
Bio/Seq.py:1426:1: D400 First line should end with a period
Bio/Seq.py:2052:1: D202 No blank lines allowed after function docstring
Bio/Seq.py:2052:1: D400 First line should end with a period
Bio/Seq.py:2070:1: D202 No blank lines allowed after function docstring
Bio/Seq.py:2070:1: D400 First line should end with a period
Bio/Seq.py:2088:1: D202 No blank lines allowed after function docstring
Bio/Seq.py:2088:1: D400 First line should end with a period
```

I've made a comment on the ``MutableSeq`` implementation - do you think working from the ``self.data`` array would be more efficient?

Have you tried using the BioSQL wrapper's ``DBSeq`` object (which subclasses ``Seq``)? My expectation is it will break - probably it needs to create a ``Seq`` object due to the nature of the ``DBSeq`` object. Tests for this would best be added to class ``SeqInterfaceTest`` in ``Tests/common_BioSQL.py`` - you should be able to test this via ``Tests/test_BioSQL_sqlite3.py`` easily (setting up the other BioSQL back-ends is more trouble).

chris-rands commented:
Thanks Peter, I corrected the style points and I totally agree about using `self.data` for the `MutableSeq` methods, did I implement it as you intended?

I am not familiar with BioSQL I'm afraid, can you provide more details on what you you would like there?

peterjc commented:
Your ``MutableSeq`` change looks great.

For BioSQL, I am picturing something like this in the tests (untested):

```python
class SeqInterfaceTest(unittest.TestCase):
...

    def test_seq(self):
        """Make sure Seqs from BioSQL implement the right interface."""
        test_seq = self.item.seq
        alphabet = test_seq.alphabet
        ...
        tripled = test_seq * 3  # check DBSeq.__mul__
        self.assertTrue(isinstance(tripled, Seq))
        self.assertEqual(tripled, str(test_seq) * 3)
        self.assertEqual(tripled.alphabet, alphabet)
        tripled = 3 * test_seq 3  # check DBSeq.__rmul__
        self.assertTrue(isinstance(tripled, Seq))
        self.assertEqual(tripled, str(test_seq) * 3)
        self.assertEqual(tripled.alphabet, alphabet)
        # Something similar for __imul__ as well...
```

(You could define a new test method if you like, expanding the existing ``test_seq`` is my being a bit lazy, but hopefully giving a consistent example.)

chris-rands commented:
Thanks, I can implement those tests, and in terms of the implementation in `BioSQL/BioSeq.py` itself, can I do something similar to `__add__` and `__radd__`, like:

    def __mull__(self, other):
        """Multiply sequence by an integer.

        The sequence is first converted to a Seq object before multiplication.
        The returned object is a Seq object, not a DBSeq object.
        """
        # Let the Seq object deal with the alphabet issues etc
        return self.toseq() * other

     def __rmull__(self, other):
        """Multiply integer by a sequence.

        The sequence is first converted to a Seq object before multiplication.
        The returned object is a Seq object, not a DBSeq object.
        """
        # Let the Seq object deal with the alphabet issues etc
        return other * self.toseq()

     def __imull__(self, other):
        """Multiply sequence by integer in-place.

        The sequence is first converted to a Seq object before multiplication.
        The returned object is a Seq object, not a DBSeq object.
        """
        # Let the Seq object deal with the alphabet issues etc
        return self.toseq() * other

As I said I don't use BioSQL, so I really don't know if this is appropriate or not? Thanks for your patience

peterjc commented:
That looks perfect for the BioSQL changes, yes please.

peterjc commented:
Just needs the BioSQL tests added, and the code side of this should be ready.

Would you also like to write a couple of lines about this for the ``NEWS.rst`` file (and add yourself to the contributors to the upcoming release)?

chris-rands commented:
Thanks Peter, I made a few mistakes (now corrected hopefully) and I've added tests but they are failing on my system at least, it looks like the `DBSeq` methods are not overiding thse `Seq` methods correctly, what am I missing? 

```
ERROR: test_multiplication (common_BioSQL.SeqInterfaceTest)
Check can multiply DBSeq objects by integers.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/cegg/rands/BIOPYTHON/biopython/Tests/common_BioSQL.py", line 543, in test_multiplication
    tripled = test_seq * 3
  File "/home/cegg/rands/.local/lib/python3.6/site-packages/Bio/Seq.py", line 360, in __mul__
    return self.__class__(str(self) * other, self.alphabet)
TypeError: __init__() missing 3 required positional arguments: 'alphabet', 'start', and 'length'
```

peterjc commented on Tests/common_BioSQL.py:
```diff
@@ -536,6 +536,28 @@ def test_addition(self):
+        alphabet = test_seq.alphabet
+        tripled = test_seq * 3
+        # Test DBSeq.__mul__
+        self.assertTrue(isinstance(tripled, Seq))
```
We should probably add a check that this is not a ``DBSeq``, i.e. that is it only a ``Seq``:

```python
self.assertFalse(isinstance(tripled, DBSeq))
```

(And repeat this later on)

peterjc commented:
If you are comfortable doing a git rebase onto the current master, that'd be great. Otherwise I'll deal with the (minor) merge conflict in the NEWS file.

peterjc commented:
Do you want to be on the list of contributors for Biopython 1.72?

chris-rands commented:
Thanks Peter, you are correct, sorry, I made a mistake with the testing, now all tests pass on my system when I run `test_BioSQL_sqlite3.py`.

I am already listed in the contributors because of a previous pull request. Are you able to resolve the conflict easily?

peterjc commented:
The merge conflict is trivial - two new news entries, keep both. Git can't decide on its own which order to put them in ;)

peterjc commented:
(I'm going to wait for the penultimate TravisCI and AppVeyor tests to finish)

peterjc commented:
Rebased to deal with the merge conflict, and applied to master - thank you!

chris-rands commented:
Great, thanks!
